File: system/loggerd/encoder/encoder.cc
#include "system/loggerd/encoder/encoder.h"

VideoEncoder::VideoEncoder(const EncoderInfo &encoder_info, int in_width, int in_height)
    : encoder_info(encoder_info), in_width(in_width), in_height(in_height) {

  out_width = encoder_info.frame_width > 0 ? encoder_info.frame_width : in_width;
  out_height = encoder_info.frame_height > 0 ? encoder_info.frame_height : in_height;


  std::vector pubs = {encoder_info.publish_name};
  if (encoder_info.thumbnail_name != NULL) {
    pubs.push_back(encoder_info.thumbnail_name);
  }
  pm.reset(new PubMaster(pubs));
}

void VideoEncoder::publisher_publish(int segment_num, uint32_t idx, VisionIpcBufExtra &extra,
                                     unsigned int flags, kj::ArrayPtr<capnp::byte> header, kj::ArrayPtr<capnp::byte> dat) {
  // broadcast packet
  MessageBuilder msg;
  auto event = msg.initEvent(true);
  auto edat = (event.*(encoder_info.init_encode_data_func))();
  auto edata = edat.initIdx();
  struct timespec ts;
  timespec_get(&ts, TIME_UTC);
  edat.setUnixTimestampNanos((uint64_t)ts.tv_sec*1000000000 + ts.tv_nsec);
  edata.setFrameId(extra.frame_id);
  edata.setTimestampSof(extra.timestamp_sof);
  edata.setTimestampEof(extra.timestamp_eof);
  edata.setType(encoder_info.encode_type);
  edata.setEncodeId(cnt++);
  edata.setSegmentNum(segment_num);
  edata.setSegmentId(idx);
  edata.setFlags(flags);
  edata.setLen(dat.size());
  edat.setData(dat);
  edat.setWidth(out_width);
  edat.setHeight(out_height);
  if (flags & V4L2_BUF_FLAG_KEYFRAME) edat.setHeader(header);

  uint32_t bytes_size = capnp::computeSerializedSizeInWords(msg) * sizeof(capnp::word);
  if (msg_cache.size() < bytes_size) {
    msg_cache.resize(bytes_size);
  }
  kj::ArrayOutputStream output_stream(kj::ArrayPtr<capnp::byte>(msg_cache.data(), bytes_size));
  capnp::writeMessage(output_stream, msg);
  pm->send(encoder_info.publish_name, msg_cache.data(), bytes_size);

  // Publish keyframe thumbnail
  if ((flags & V4L2_BUF_FLAG_KEYFRAME) && encoder_info.thumbnail_name != NULL) {
    MessageBuilder tm;
    auto thumbnail = tm.initEvent().initThumbnail();
    thumbnail.setFrameId(extra.frame_id);
    thumbnail.setTimestampEof(extra.timestamp_eof);
    thumbnail.setThumbnail(dat);
    thumbnail.setEncoding(cereal::Thumbnail::Encoding::KEYFRAME);
    pm->send(encoder_info.thumbnail_name, tm);
  }
}

File: system/loggerd/encoder/encoder.py
from dataclasses import dataclass

import cereal.messaging as messaging


@dataclass
class EncoderSettings:
  encode_type: str
  bitrate: int
  gop_size: int
  b_frames: int = 0


class VideoEncoder:
  def __init__(self, encoder_info, in_width: int, in_height: int):
    self.encoder_info = encoder_info
    self.in_width = in_width
    self.in_height = in_height
    self.out_width = encoder_info.frame_width if encoder_info.frame_width > 0 else in_width
    self.out_height = encoder_info.frame_height if encoder_info.frame_height > 0 else in_height
    self.pm = messaging.PubMaster([encoder_info.publish_name])
    self.encode_cnt = 0
    self.segment_num = -1
    self.counter = 0

  def open(self) -> None:
    raise NotImplementedError

  def close(self) -> None:
    raise NotImplementedError

  def encode_frame(self, buf, frame_id: int, timestamp_sof: int, timestamp_eof: int) -> int:
    raise NotImplementedError
